src/compiler_rt.zig

```diff
@@ -0,0 +1,186 @@
+const std = @import("std");
+const builtin = @import("builtin");
+const Allocator = std.mem.Allocator;
+const mem = std.mem;
+const tracy = @import("tracy.zig");
+const trace = tracy.trace;
+
+const Compilation = @import("Compilation.zig");
+const CRTFile = Compilation.CRTFile;
+const LinkObject = Compilation.LinkObject;
+const Package = @import("Package.zig");
+
+pub const CompilerRtLib = struct {
+    crt_object_files: [sources.len]?CRTFile = undefined,
+    crt_lib_file: ?CRTFile = null,
+
+    pub fn deinit(crt_lib: *CompilerRtLib, gpa: Allocator) void {
+        for (crt_lib.crt_object_files) |*crt_file| {
+            if (crt_file.*) |*cf| {
+                cf.deinit(gpa);
+            }
+        }
+        if (crt_lib.crt_lib_file) |*crt_file| {
+            crt_file.deinit(gpa);
+        }
+    }
+};
+
+pub fn buildCompilerRtLib(comp: *Compilation, compiler_rt_lib: *CompilerRtLib) !void {
+    const tracy_trace = trace(@src());
+    defer tracy_trace.end();
+
+    var progress: std.Progress = .{ .dont_print_on_dumb = true };
+    var progress_node = progress.start("Compile Compiler-RT", sources.len + 1);
+    defer progress_node.end();
+    if (comp.color == .off) progress.terminal = null;
+
+    progress_node.activate();
+
+    var link_objects: [sources.len]LinkObject = undefined;
+    for (sources) |source, i| {
+        var obj_progress_node = progress_node.start(source, 0);
+        obj_progress_node.activate();
+        defer obj_progress_node.end();
+
+        try comp.buildOutputFromZig(source, .Obj, &compiler_rt_lib.crt_object_files[i], .compiler_rt);
+        link_objects[i] = .{
+            .path = compiler_rt_lib.crt_object_files[i].?.full_object_path,
+            .must_link = true,
+        };
+    }
+
+    const root_name = "compiler_rt";
+
+    var lib_progress_node = progress_node.start(root_name, 0);
+    lib_progress_node.activate();
+    defer lib_progress_node.end();
+
+    const target = comp.getTarget();
+    const basename = try std.zig.binNameAlloc(comp.gpa, .{
+        .root_name = root_name,
+        .target = target,
+        .output_mode = .Lib,
+    });
+    errdefer comp.gpa.free(basename);
+
+    // TODO: This is extracted into a local variable to work around a stage1 miscompilation.
+    const emit_bin = Compilation.EmitLoc{
+        .directory = null, // Put it in the cache directory.
+        .basename = basename,
+    };
+    const sub_compilation = try Compilation.create(comp.gpa, .{
+        .local_cache_directory = comp.global_cache_directory,
+        .global_cache_directory = comp.global_cache_directory,
+        .zig_lib_directory = comp.zig_lib_directory,
+        .cache_mode = .whole,
+        .target = target,
+        .root_name = root_name,
+        .main_pkg = null,
+        .output_mode = .Lib,
+        .link_mode = .Static,
+        .function_sections = true,
```

Comment:
do we still want this now that all the functions are in separate object files?

Reply:
I think so. Here's why I think so. Since we bundle up a couple related symbols together in one object file, we probably want this on for the linker to GC unused symbol versions.